src/MichiPointsMinter.sol

```diff
@@ -0,0 +1,170 @@
+
+    /// @notice error when 0 tokens are requested to be minted
+    error InvalidAmount(uint256 amount);
+
+    /// @notice error when request can already been fulfilled
+    error RequestAlreadyFulfilled(uint256 chainId, uint256 requestId);
+
+    /// @notice error when tokenize fee is set too high
+    error InvalidTokenizeFee(uint256 tokenizeFee);
+
+    /// @notice error when approving an already approved token
+    error TokenizedPointAlreadyApproved(address tokenizedPoint);
+
+    /// @notice error when removing an unapproved token
+    error TokenizedPointNotApproved(address tokenizedPoint);
+
+    /// @notice event when a tokenize point request has been completed
+    event RequestFulfilled(
+        address receier, address[] tokenizedPoints, uint256 chainId, uint256 requestId, uint256[] amounts
+    );
+
+    /// @dev Constructor for MichiPointsMinter contract
+    /// @param feeReceiver_ address of the fee receiver
+    /// @param tokenizeFee_ tokenize fee
+    /// @param precision_ precision denominator
+    constructor(address feeReceiver_, uint256 tokenizeFee_, uint256 precision_) {
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _setupRole(MINTER_ROLE, msg.sender);
+
+        feeReceiver = feeReceiver_;
+        tokenizeFee = tokenizeFee_;
+        precision = precision_;
+    }
+
+    /// @dev Initiate tokenized points minting
+    /// @param receiver address receiving the minted tokens
+    /// @param tokensToMint array of tokens to mint
+    /// @param amounts array of amounts to mint
+    /// @param chainId chain that request was made on
+    /// @param requestId id of request
+    function mintTokenizedPoints(
+        address receiver,
+        address[] calldata tokensToMint,
+        uint256[] calldata amounts,
+        uint256 chainId,
+        uint256 requestId
+    ) external onlyRole(MINTER_ROLE) {
+        if (tokensToMint.length != amounts.length) revert ArrayLengthMismatch();
+
+        for (uint256 i = 0; i < tokensToMint.length; i++) {
+            if (!approvedTokenizedPoints[tokensToMint[i]]) revert UnapprovedToken(tokensToMint[i]);
+            if (amounts[i] == 0) revert InvalidAmount(amounts[i]);
+
+            if (tokenizeFee == 0) {
+                ITokenizedPointERC20(tokensToMint[i]).mint(receiver, amounts[i]);
+                userAmountMintedByTokenizedPoint[tokensToMint[i]] += amounts[i];
+            } else {
+                uint256 fee = amounts[i] * tokenizeFee / precision;
+                uint256 amountAfterFees = amounts[i] - fee;
+
+                ITokenizedPointERC20(tokensToMint[i]).mint(receiver, amountAfterFees);
+                ITokenizedPointERC20(tokensToMint[i]).mint(feeReceiver, fee);
+
+                userAmountMintedByTokenizedPoint[tokensToMint[i]] += amountAfterFees;
+                feesByTokenizedPoint[tokensToMint[i]] += fee;
+            }
+        }
+        FulfilledRequest storage fulfilledRequest = chainToRequestId[chainId][requestId];
```

Comment:
might be good to add the requestor contract address as well,... chainToRequestId[chainId][requestor][requestId]